[WIP] [IMP] allow using sudo in the OCA runbot

gurneyalex opened the pull request:
Current situation; the install section of travis is mapped to
instructions running with a non root user.

However, some setups in OCA projects such as l10n-brazil, are using
sudo in their travis file to run things such as locales updates, which
fails (blocked on password prompt).

With this change, we parse the lines starting with sudo in the section
and place them in a different script, called sudo_runs. This script is executed
*after* the runs script, in a section where the user is root.

known limitation: this does not work if the commands requiring sudo and the others
must be run in a specific order. But it should be good enough for oca/l10n-brazil.

moylop260 commented:
Hi @gurneyalex 
`sudo` is not allowed for runbot instance since that it could be a risk of security.

For the particular case of brazil travis is installing locales if we request it. (I don't remember the thread but we had the same issue a long time ago)

Let me check if I can fix it using another way.

What do you think?

gurneyalex commented:
hello @moylop260 thanks for getting back on this. 

This PR is an attempt to emulate sudo during the docker container build by changing the user, but you are right there might be some security issues that I'm not aware of. 

If you have a way to setup locale in the runbot containers, I'm sure this would be welcome by l10-brazil and @rvalyi

rvalyi commented:
Hello @gurneyalex  and @moylop260 thank you for your attention on this one.
However, as dicsussed in https://github.com/OCA/l10n-brazil/pull/726 this is only for Runbot on legacy v8 and v10 series. As for v12 and upward we deprecated the legacy electronic invoicing lib that was forcing us to do these locale trickery (pysped) so to speed up the builds, we just decided to even remove these sudo scripts on v12. However the Runbot doesn't run either with no log and no clue as reported here https://github.com/OCA/l10n-brazil/pull/726
do you guys have an idea what is going on?

gurneyalex commented:
@rvalyi the runbot was still having the sudo in travis.yml to cope with and was running with a tweaked docker image of my own, which had its own problems. 

I just removed the sudo parts of setup and reverted to the normal runbot base docker image, and 

![image](https://user-images.githubusercontent.com/2032807/66431093-a123b800-ea1b-11e9-962b-dee1077bb882.png)

rvalyi commented:
@gurneyalex ah thanks! was thinking about that kind of thing indeed. So for 12.0 and 13.0 we can go with the standard image and for previous versions let's see if this PR and a custom image fix our problems.. Thank you very much, your help is really appreciated!

gurneyalex commented:
@rvalyi yes.

For previous versions I see 2 options:

* easy: find a way to ignore the locale tweaking in the context of the runbot : this allows you to test everything but the bits depending on pysped, which could still be useful

* harder: find a way in travis2docker to run the bits we need to set up the locale as needed. 

Main issue I found is that, while I can build a new docker base image with the locale available, then travis2docker has some hard coded image names built in which are hard to override without making a new release, and depending on the image name, it will create a different Dockerfile (see [the template Dockerfile|https://github.com/Vauxoo/travis2docker/blob/3a88783c8a1c65e016c6f20407aa4a92166cb373/src/travis2docker/templates/Dockerfile#L77]

I wish we had a more configurable way of setting things up here.

rvalyi commented:
@gurneyalex honnestly now that the issue is identified (we had no clue before you told us), I think this shouldn't be too hard to have a fallback system in our legacy electronic invoicing lib pysped where it would just use the wrong en_US locale in the XML electronic invoices if pt_BR is not available. The XML on runbot would not be valid, but we really don't care. Prior to 12.0 we didn't even test the generated XML values. And in v12.0 we do way more checks but not using pysped anymore so we are safe. cc @mileo @renatonlima @ananiasfilho @mbcosta

gurneyalex commented:
runbot can be unforgiving when it does not work as expected at all... It took me a while to figure out why the builds were killed...

moylop260 commented:
@ruiztulio 
Could you pre-install in image used for runbot Brazil locales?

moylop260 commented:
@gurneyalex @rvalyi 
I found the thread about:
 - https://github.com/travis-ci/travis-ci/issues/4416#issuecomment-278163295

The solution is:
```yml
addons:
    apt:
        packages:
            - language-pack-br
```

moylop260 commented:
I just have created the following PR:
 - https://github.com/OCA/l10n-brazil/pull/732

moylop260 commented:
The runbot build using apt-package language-pack-br is green:
 - http://runbot.odoo-community.org/runbot/build/3389995

gurneyalex commented:
Thanks, I'm closing this PR.
